Have natural=water follow waterway=* on checking for layer.

danfos opened the pull request:
First try in #937 but could not update that PR anymore because I did rebase.

Did analyze #934 further and saw at least one problem was for waterway=* + tunnel=culvert + level=-1. The problem here is the key level, that should be layer instead. This patch adds details to clarify that.

The other problem is for natural=water + tunnel=culvert + level=-1, this patch makes natural=water follow waterway=* on checking layer as that code has already all the good things.

Did check [natural=*](https://wiki.openstreetmap.org/wiki/Key:natural) but only for natural-water layer=* makes sense to me.

frodrigo commented:
Ok, for me.
But tests fails:
```
    def way(self, data, tags, nds):

        if tags.get(u"layer") and tags.get(u"layer") != "0":

            layer = tags.get(u"layer")

            if tags.get(u"landuse"):

                return {"class": 41101, "subclass": 0}

            elif tags.get(u"natural") and layer[0] == '-' and tags.get(u"natural") != "water":

                return {"class": 41102, "subclass": 0}

            elif tags.get(u"highway") and tags.get(u"highway") != "steps" and (not tags.get(u"indoor") or tags.get(u"indoor") == "no"):

                if layer[0] == "-" and (not tags.get(u"tunnel") or tags.get(u"tunnel") == "no"):

                    return {"class": 41104 if len(nds) > 3 else 41103, "subclass": 0}

                elif layer[0] != "-" and (not tags.get(u"bridge") or tags.get(u"bridge") == "no") and tags.get(u"man_made") != "pier":

                    if len(nds) > 3:

                        return {"class": 41106, "subclass": 0, "fix": {"-": ["layer"]}}

                    else:

                        return {"class": 41105, "subclass": 0, "fix": {"+": {"bridge": "yes"}}}

            elif tags.get(u"waterway") or ((tags.get(u"natural") and tags.get(u"natural") == "water")):

                if layer[0] == "-" and (not tags.get(u"tunnel") or tags.get(u"tunnel") == "no"):

                    if len(nds) > 3:

                        return {"class": 41108, "subclass": 0, "fix": {"-": ["layer"]}}

                    else:

                        return {"class": 41107, "subclass": 0}

                elif layer[0] != "-" and (not tags.get(u"bridge") or tags.get(u"bridge") == "no"):

>                   return {"class": 41100 if len(nds) > 3 else 41109, "subclass": 0}

E                   TypeError: object of type 'NoneType' has no len()

plugins/TagRemove_Layer.py:72: TypeError
```

danfos commented:
Yes, saw the error this morning but had to go to work. Let me see if I can fix it.

danfos commented:
Fixed one problem I think but there are many others that I likely can also fix but that will take time.
I am surprised that changing some text requires many updates in the test data.

frodrigo commented:
Forget about the test data and version error, we need to change this.
Just look at the error I pointed.

danfos commented:
Had the fix ready for the test data so pushed it.

I did not see the "TypeError: object of type 'NoneType' has no len()" in the Travis log, how was it triggered?

The error is for line 72:

return {"class": 41100 if len(nds) > 3 else 41109, "subclass": 0}

but that line was not touched by the commit.

The problem seems to be that the nds variable has the value None but that is an function argument, so it is the calling function doing that.

frodrigo commented:
Yes the test set no nodes https://github.com/osm-fr/osmose-backend/blob/999e0a672614b089cf755e7169d385f6534aded0/plugins/TagRemove_Layer.py#L84
But previously this part of code was not called by the test (it is bad, but it is not the point). But the change in the previous if change the flow of process to this part, and it is not want we want. It is due to the add of ` and tags.get(u"natural") != "water"`

danfos commented:
Thanks for the hint!

Took me some time to understand nds means nodes so normally this function is called with a list of nodes of the way.

Fixed the natural test and added waterway/water tests, let's see.

danfos commented:
Okay, looks to me the problems with the changes an sich are gone, but what now seems to fail is the checksum, as the file changed the checksum changed?!

Just a matter of updating the magic number in modules/SourceVersion.py?

frodrigo commented:
No, about source version it is ok, is liked to an other issue.

frodrigo commented:
Thank. Merged, but not deployed, yet.

danfos commented:
Thanks, again some issues down on the list of objects I have recently edited with Osmose warnings once deployed.
